src/rust/engine/fs/store/src/remote.rs

```diff
@@ -127,94 +201,115 @@ impl ByteStore {
-            futures::future::ready(None)
-          } else {
-            let mut req = bazel_protos::bytestream::WriteRequest::new();
-            req.set_resource_name(resource_name.clone());
-            req.set_write_offset(offset as i64);
-            let next_offset = min(offset + chunk_size_bytes, bytes.len());
-            req.set_finish_write(next_offset == bytes.len());
-            req.set_data(Bytes::from(&bytes[offset..next_offset]));
-            futures::future::ready(Some((
-              Ok((req, grpcio::WriteFlags::default())),
-              (next_offset, true),
-            )))
-          }
-        });
+      let chunk_size_bytes = store.chunk_size_bytes;
 
-        sender.send_all(&mut stream).await.or_else(move |e| {
-          match e {
-            // Some implementations of the remote execution API early-return if the blob has
-            // been concurrently uploaded by another client. In this case, they return a
-            // WriteResponse with a committed_size equal to the digest's entire size before
-            // closing the stream.
-            // Because the server then closes the stream, the client gets an RpcFinished
-            // error in this case. We ignore this, and will later on verify that the
-            // committed_size we received from the server is equal to the expected one. If
-            // these are not equal, the upload will be considered a failure at that point.
-            // Whether this type of response will become part of the official API is up for
-            // discussion: see
-            // https://groups.google.com/d/topic/remote-execution-apis/NXUe3ItCw68/discussion.
-            grpcio::Error::RpcFinished(None) => Ok(()),
-            e => Err(format!(
-              "Error attempting to upload digest {:?}: {:?}",
-              digest, e
-            )),
-          }
-        })?;
+      // TODO(tonic): The call into the Tonic library wants the slice to last for the 'static
+      // lifetime but the slice passed into this method generally points into the shared memory
+      // of the LMDB store which is on the other side of the FFI boundary.
+      let bytes = Bytes::copy_from_slice(bytes);
 
-        // The gRPC library cancels streams on drop; closes must be explicit. Not closing
-        // the stream caused the BuildGrid CAS server to generate errors on writes
-        // when the stream was cancelled because it was not closed explicitly.
-        sender.close().await.or_else(|err| {
-          match err {
-            // Some servers (e.g., RBE) may have already closed the stream for the early
-            // return reason identified previously. Treat this condition as a successful close.
-            grpcio::Error::RpcFinished(None) => Ok(()),
-            e => Err(format!(
-              "Error from server when uploading digest {:?}: {:?}",
-              digest, e
-            )),
-          }
+      // let stream = async_stream::stream! {
```

Comment:
Left this commented-out code in because it may be a preferred solution instead of using `unfold`. Thoughts?

Reply:
Might do as a follow-up. Going to leave the existing `unfold`-based version for now.